# Librería **NumPy** (Numerical Python)

### Librería estándar para trabajar con datos numéricos. Permite generar y manipular datos de manera eficiente. 

### Incorpora la clase **array**.

----------

#### arrays (numpy) vs listas (python nativo)

Las listas son buenos contenedores para cualquier tipo de objeto, es fácil insertar o eliminar elementos. Sin embargo, el hecho de permitir diferentes tipos de elementos requiere espacio en la memoria para cada uno. Además, las operaciones entre elementos son limitadas.

Los arrays son contenedores de **un solo tipo de variables**, ocupan menos memoria y se vuelve mas eficiente su manipulación. Su estructura facilia las operaciones matemáticas entre elementos.

----------

In [1]:
import numpy as np

a = np.array([1,2,3,4])

lista = ['a','b','c']
b = np.array(lista)

print(type(a))
print(type(b))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [2]:
print(a.dtype)
print(b.dtype)

int64
<U1


In [3]:
c = np.array(['c',2.9,3,4])
print(c.dtype)

<U32


In [4]:
c

array(['c', '2.9', '3', '4'], dtype='<U32')

In [5]:
# array de una dimension: vector

d1 = np.array([1,2,3,4])
print('1D ', d1)


1D  [1 2 3 4]


In [6]:
# array de dos dimensiones: matriz
d2 = np.array([[1,2,3,4],[1,2,3,4]]) #2x4
print('2D ', d2)

2D  [[1 2 3 4]
 [1 2 3 4]]


In [7]:

# array de tres dimensiones: cubo
d3 = np.array([[[1,2,3,4],[1,2,3,4]],[[1,2,3,4],[1,2,3,4]]]) #2x4x2
print('3D ', d3)

3D  [[[1 2 3 4]
  [1 2 3 4]]

 [[1 2 3 4]
  [1 2 3 4]]]


In [8]:
d3[0][1][0]

1

In [9]:
lista=[1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7]
lista[3:9:2] # elementos con indice del 3 al 9 con saltos de dos

[4, 6, 8]

## Atributos de un array

Características

In [10]:
d3.dtype

dtype('int64')

In [11]:
d3_float = d3.astype('float')
d3_float.dtype

dtype('float64')

In [12]:
d3.ndim # numero de dimensiones

3

In [13]:
d3.shape

(2, 2, 4)

In [14]:
d3.size # numero de elementos

16

## Funciones generadoras de arrays

In [15]:
np.arange(3,20,1.5) #(inicio,fin, paso)

array([ 3. ,  4.5,  6. ,  7.5,  9. , 10.5, 12. , 13.5, 15. , 16.5, 18. ,
       19.5])

In [16]:
np.linspace(3,20,25) #(inicio,fin, numero de valores en el intervalo)

array([ 3.        ,  3.70833333,  4.41666667,  5.125     ,  5.83333333,
        6.54166667,  7.25      ,  7.95833333,  8.66666667,  9.375     ,
       10.08333333, 10.79166667, 11.5       , 12.20833333, 12.91666667,
       13.625     , 14.33333333, 15.04166667, 15.75      , 16.45833333,
       17.16666667, 17.875     , 18.58333333, 19.29166667, 20.        ])

In [17]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [18]:
np.ones(10)*23

array([23., 23., 23., 23., 23., 23., 23., 23., 23., 23.])

In [19]:
np.zeros((5,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [20]:
np.identity(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

#### Números aleatorios

* rand :  entre 0 y 1

* randn : distribución normal estándar

* randint : enteros



In [21]:
np.random.rand(5) #0-1

array([0.17281117, 0.9795514 , 0.57026081, 0.72106321, 0.69111926])

In [22]:
np.random.rand(20)*4 + 4 #cualquier rango

array([5.19549357, 7.63792198, 6.65700323, 5.12726326, 7.78837506,
       6.45518467, 7.19215061, 4.49359774, 4.48291616, 5.48197317,
       4.75955594, 5.38058208, 5.12585468, 5.95172545, 4.743479  ,
       5.85772661, 6.34421945, 7.02021068, 6.33121768, 4.62702633])

In [23]:
np.random.randn(3) # distribucion normal

array([ 0.02494767, -0.36442551,  0.20036242])

In [24]:
np.random.randn(3,3) # distribucion normal

array([[ 2.06044711,  1.00247255, -0.67222007],
       [ 0.69297217, -0.29951467, -2.00711422],
       [-1.87241488, -0.06264069, -0.62756137]])

In [25]:
np.random.randint(20,80,5) #(inicio,fin,cuantos)

array([29, 44, 73, 34, 50])

## Manipulación de arrays

[Array manipulation routines](https://numpy.org/doc/stable/reference/routines.array-manipulation.html)

append, insert, concat, reshape

In [26]:
print(d3)
print(d3.shape)

[[[1 2 3 4]
  [1 2 3 4]]

 [[1 2 3 4]
  [1 2 3 4]]]
(2, 2, 4)


In [27]:
d3_a_d2 = d3.reshape(4,4)
print(d3_a_d2)
print(d3_a_d2.shape)

[[1 2 3 4]
 [1 2 3 4]
 [1 2 3 4]
 [1 2 3 4]]
(4, 4)


In [28]:
x = np.array([2, 4, -4, 9, -1])

np.concatenate((x, [1000, 2222], d3_a_d2[0], d3_a_d2[:,3]))

array([   2,    4,   -4,    9,   -1, 1000, 2222,    1,    2,    3,    4,
          4,    4,    4,    4])

In [29]:
np.append(x, 100) 

array([  2,   4,  -4,   9,  -1, 100])

In [30]:
np.append(x, np.nan)

array([ 2.,  4., -4.,  9., -1., nan])

In [31]:
np.insert(x, 2, 200) # insertar 200 en la posicion con indice 2


array([  2,   4, 200,  -4,   9,  -1])

In [32]:
x = np.arange(20)
x

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])

In [33]:
np.insert(x, [2, 4, -1], [222, 444, -111])

array([   0,    1,  222,    2,    3,  444,    4,    5,    6,    7,    8,
          9,   10,   11,   12,   13,   14,   15,   16,   17,   18, -111,
         19])

In [34]:
d2[0]

array([1, 2, 3, 4])

In [35]:
np.insert(d2[0], 0,5)

array([5, 1, 2, 3, 4])

In [36]:
d2

array([[1, 2, 3, 4],
       [1, 2, 3, 4]])

In [37]:
np.append(d2, [[7, 8, 9, 10]], axis=0)

array([[ 1,  2,  3,  4],
       [ 1,  2,  3,  4],
       [ 7,  8,  9, 10]])

In [38]:
d2

array([[1, 2, 3, 4],
       [1, 2, 3, 4]])

In [39]:
#np.append(d2, [[7, 8]], axis=1) # error

In [40]:
a_to_add = np.reshape([7,8], (2,1)) # array de dos filas y una columna

np.append(d2, a_to_add, axis=1)

array([[1, 2, 3, 4, 7],
       [1, 2, 3, 4, 8]])

In [41]:
np.append(d2, [[7],[8]], axis=1)

array([[1, 2, 3, 4, 7],
       [1, 2, 3, 4, 8]])

In [42]:
np.insert(d2, 1, 10, axis=1)

array([[ 1, 10,  2,  3,  4],
       [ 1, 10,  2,  3,  4]])

In [43]:
np.insert(d2, 1, [[7,8]], axis=1)

array([[1, 7, 2, 3, 4],
       [1, 8, 2, 3, 4]])

In [44]:
np.insert(d2, 1, 10, axis=0)

array([[ 1,  2,  3,  4],
       [10, 10, 10, 10],
       [ 1,  2,  3,  4]])

## Operaciones con arrays

Operaciones elemento a elemento

In [45]:
x = np.array([2.0, 4.6, -4.6, 9.3, -1.2])

y = np.arange(len(x))
y

array([0, 1, 2, 3, 4])

In [46]:
y*y

array([ 0,  1,  4,  9, 16])

In [47]:
y*y + 2*x

array([ 4. , 10.2, -5.2, 27.6, 13.6])

In [48]:
z = y*y + 2*x
z**3 - 5

array([   59.   ,  1056.208,  -145.608, 21019.576,  2510.456])

In [49]:
np.floor(z) #redondea hacia abajo
np.ceil(z) #redondea hacia arriba
w = np.absolute(z)
np.sqrt(w)
np.min(z)
np.max(z)
np.mean(z)
np.median(z)
np.sum(z)

50.2

In [50]:
z

array([ 4. , 10.2, -5.2, 27.6, 13.6])

In [51]:
z*0.1234

array([ 0.4936 ,  1.25868, -0.64168,  3.40584,  1.67824])

In [52]:
np.round(z*0.1234, 1)

array([ 0.5,  1.3, -0.6,  3.4,  1.7])

In [53]:
np.random.seed(2)

m1 = np.random.randint(10,30,(4,5))
m1

array([[18, 25, 23, 18, 21],
       [28, 21, 18, 17, 12],
       [27, 21, 25, 15, 17],
       [13, 16, 14, 20, 21]])

In [54]:
m2 = np.random.randint(0,5,(4,5))
m2

array([[3, 2, 1, 2, 4],
       [3, 0, 4, 3, 1],
       [2, 0, 4, 4, 2],
       [4, 2, 1, 0, 2]])

In [55]:
m1 + m2 

array([[21, 27, 24, 20, 25],
       [31, 21, 22, 20, 13],
       [29, 21, 29, 19, 19],
       [17, 18, 15, 20, 23]])

## Filtros y máscaras

array[condicion]

el modulo [mask](https://numpy.org/doc/stable/reference/maskedarray.generic.html) permite la generación y manipulación de máscaras

In [56]:
x = np.arange(5, 10)
x

array([5, 6, 7, 8, 9])

In [57]:
b = np.array([True, False, False, True, False])
b

array([ True, False, False,  True, False])

In [58]:
b = np.array([True, False, False, True, False], dtype=bool) # buenas practicas
b

array([ True, False, False,  True, False])

In [59]:
x[b]

array([5, 8])

In [60]:
x>7

array([False, False, False,  True,  True])

In [61]:
x%2 == 1

array([ True, False,  True, False,  True])

In [62]:
m2 == 0 # mascara

array([[False, False, False, False, False],
       [False,  True, False, False, False],
       [False,  True, False, False, False],
       [False, False, False,  True, False]])

In [63]:
mask = m2 == 0 # mascara
mask

array([[False, False, False, False, False],
       [False,  True, False, False, False],
       [False,  True, False, False, False],
       [False, False, False,  True, False]])

In [64]:
mask = m2 >= 2
mask

array([[ True,  True, False,  True,  True],
       [ True, False,  True,  True, False],
       [ True, False,  True,  True,  True],
       [ True,  True, False, False,  True]])

In [65]:
m2[mask]

array([3, 2, 2, 4, 3, 4, 3, 2, 4, 4, 2, 4, 2, 2])

In [66]:
m2*mask # los False los pone como 0

array([[3, 2, 0, 2, 4],
       [3, 0, 4, 3, 0],
       [2, 0, 4, 4, 2],
       [4, 2, 0, 0, 2]])

In [67]:
import numpy.ma as ma

x = np.arange(5, 10)

ma.masked_array(x, mask=[1, 1, 0, 1, 0])

masked_array(data=[--, --, 7, --, 9],
             mask=[ True,  True, False,  True, False],
       fill_value=999999)

In [68]:
x_masked = ma.masked_array(x, mask=[1, 1, 0, 1, 0])

x_masked.mean()

8.0

In [69]:
ma.masked_array(m1, mask = m1%2 == 0)

masked_array(
  data=[[--, 25, 23, --, 21],
        [--, 21, --, 17, --],
        [27, 21, 25, 15, 17],
        [13, --, --, --, 21]],
  mask=[[ True, False, False,  True, False],
        [ True, False,  True, False,  True],
        [False, False, False, False, False],
        [False,  True,  True,  True, False]],
  fill_value=999999)

In [70]:
x = [-9999.,1.,-9999.,3.,4.]
mx = ma.masked_values (x, -9999.)
mx

masked_array(data=[--, 1.0, --, 3.0, 4.0],
             mask=[ True, False,  True, False, False],
       fill_value=-9999.0)

In [71]:
mx.filled(mx.mean()) #rellenar los valores vacios

array([2.66666667, 1.        , 2.66666667, 3.        , 4.        ])

## Algebra lineal

submódulo **linalg** para operaciones algebraicas de vectores y matrices.

Producto punto

In [110]:
# producto escalar de dos vectores
a = np.array([1, 2, 3, 4])
b = np.array([1, 0, 1, 2])

a.dot(b)

12

In [111]:
a @ b

12

In [112]:
A = np.array([[1,2,3], [4,5,6], [7,8,9]])
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [113]:
A*A # elemento por elemento

array([[ 1,  4,  9],
       [16, 25, 36],
       [49, 64, 81]])

In [114]:
A.dot(A) #producto matricial

array([[ 30,  36,  42],
       [ 66,  81,  96],
       [102, 126, 150]])

In [77]:
A_matriz = np.mat(A) 

In [115]:
A_matriz*A_matriz  # producto matricial (filas x columnas), no elemento a elemento como con arrays

matrix([[ 30,  36,  42],
        [ 66,  81,  96],
        [102, 126, 150]])

In [116]:
A.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [117]:
A_matriz.T 

matrix([[1, 4, 7],
        [2, 5, 8],
        [3, 6, 9]])

In [118]:
A_matriz.diagonal()

matrix([[1, 5, 9]])

In [119]:
A_matriz.trace() # traza

matrix([[15]])

In [120]:
np.linalg.det(A_matriz)  

-9.51619735392994e-16

In [121]:
a = np.array([[1, 2], [3, 4]])
np.linalg.inv(a) # A * Ainv = identidad

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [122]:
A_inv = np.linalg.inv(a)
a.dot(A_inv)

array([[1.00000000e+00, 1.11022302e-16],
       [0.00000000e+00, 1.00000000e+00]])

## Estadística

el modulo [statistics](https://numpy.org/doc/stable/reference/routines.statistics.html) tiene funciones básicas de estadística descriptiva

In [123]:
data = np.random.rand(100)

In [124]:
np.ptp(data) # rango: max-min

0.9857641154135763

In [125]:
np.percentile(data, 5) # percentiles

0.039989036874858955

In [126]:
np.median(data)
np.mean(data)
np.std(data)
np.var(data)

0.08572221169140125

In [127]:
col1 = np.random.rand(10)*3 + 10
col2 = np.random.randn(10)*3 + 10

In [128]:
data = np.concatenate((col1.reshape(10,1), col2.reshape(10,1)), axis=1)
data

array([[12.50489756, 12.79086962],
       [10.30202699,  6.31200653],
       [11.09643515,  4.86571931],
       [11.76729134,  8.36061105],
       [11.41208501, 12.35803419],
       [11.0976597 ,  7.26198032],
       [12.91164327,  8.23229754],
       [11.59715472,  8.25778103],
       [11.20838482,  6.75481533],
       [10.12373786, 12.47324737]])

In [129]:
np.min(data)

4.865719305550311

In [93]:
np.min(data, axis=0)

array([10.07915245,  5.63194921])

In [94]:
np.min(data, axis=1)

array([11.67316304,  8.604312  ,  9.87164175, 10.06306074,  5.63194921,
       10.82631361, 10.43563107, 11.5162349 ,  9.51791795, 10.48508629])

In [95]:
np.std(data, axis=0)

array([0.78959836, 2.20597933])

In [96]:
pesos = np.arange(20).reshape(10,2)

In [130]:
#promedio pesado  avg = sum(a * weights) / sum(weights)
np.average(data, axis=0, weights = pesos) 

array([11.29697937,  8.88393998])

In [131]:
np.mean(data, axis=0)

array([11.40213164,  8.76673623])

In [99]:
np.corrcoef(data[:,0], data[:,1]) # matriz de correlacion

array([[ 1.        , -0.34815402],
       [-0.34815402,  1.        ]])

In [100]:
np.histogram(data[:,1])

(array([1, 0, 0, 1, 1, 3, 0, 2, 1, 1]),
 array([ 5.63194921,  6.47071295,  7.3094767 ,  8.14824044,  8.98700418,
         9.82576793, 10.66453167, 11.50329541, 12.34205916, 13.1808229 ,
        14.01958664]))

In [101]:
np.histogram2d(data[:,0], data[:,1])

(array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 1., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 0., 1., 0.],
        [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0., 0., 1.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
        [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]),
 array([10.07915245, 10.33214979, 10.58514713, 10.83814448, 11.09114182,
        11.34413916, 11.59713651, 11.85013385, 12.10313119, 12.35612854,
        12.60912588]),
 array([ 5.63194921,  6.47071295,  7.3094767 ,  8.14824044,  8.98700418,
         9.82576793, 10.66453167, 11.50329541, 12.34205916, 13.1808229 ,
        14.01958664]))

## Lectura y escritura de documentos

el modulo [mask](https://numpy.org/doc/stable/reference/maskedarray.generic.html) permite la generación y manipulación de máscaras

In [132]:
data = np.random.rand(10,2)

In [103]:
data

array([[0.91065551, 0.81946537],
       [0.09037037, 0.35824031],
       [0.01898513, 0.45002672],
       [0.03036825, 0.27035397],
       [0.00258258, 0.43374592],
       [0.5871742 , 0.99198426],
       [0.43552082, 0.36242738],
       [0.87312734, 0.48505377],
       [0.74386203, 0.75418095],
       [0.3236832 , 0.73535139]])

In [135]:
np.savetxt('datos.txt', data, delimiter=',', fmt='%.3f', header="tiempo\t distancia")





In [136]:
data_read = np.loadtxt("datos.txt", delimiter=",")

In [137]:
# Lectura
data_read = np.loadtxt("datos.txt", delimiter=",")
tiempo = data_read[:, 0]  # primera columna
masa = data_read[:, 1]  # segunda columna

In [138]:
tiempo

array([0.143, 0.945, 0.679, 0.574, 0.133, 0.144, 0.413, 0.85 , 0.906,
       0.467])

In [139]:
t, d = np.loadtxt("datos.tsv", delimiter="\t", unpack=True) #unpack lo lee por columnas

In [140]:
t

array([0.143, 0.945, 0.679, 0.574, 0.133, 0.144, 0.413, 0.85 , 0.906,
       0.467])


-------

### Ejercicio 3:

i. Reproducir la matriz B: 

[[0.4  0.41 0.42 0.43 0.44]\
[0.45 0.46 0.47 0.48 0.49]\
[0.5  0.51 0.52 0.53 0.54]\
[0.55 0.56 0.57 0.58 0.59]\
[0.6  0.61 0.62 0.63 0.64]\
[0.65 0.66 0.67 0.68 0.69]\
[0.7  0.71 0.72 0.73 0.74]\
[0.75 0.76 0.77 0.78 0.79]]

ii. Unir B con el resultado del ejercicio anterior 2.ii para obtener una matriz de 8 lineas y 13 columnas. 

iii. Agregar una linea al final con el promedio de cada columna

iv. Agregar una columna al final con el valor maximo de cada fila

v. Guardar en un archivo.